File: backend/app/routers/vendor.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.dependencies import get_current_user
from app.models.user import User

from app.schemas.vendor import (
    VendorCreate,
    VendorUpdate,
    VendorResponse,
    VendorApprovalResponse,
)

from app.services.vendor_service import (
    create_vendor_service,
    get_all_vendors_service,
    get_vendor_by_id_service,
    update_vendor_service,
    delete_vendor_service,
)

from app.services.vendor_approval_service import (
    get_pending_vendors_service,
    approve_vendor_service,
    reject_vendor_service,
)

router = APIRouter(
    prefix="/vendors",
    tags=["Vendors"],
)


# ----------------------------
# Vendor CRUD APIs
# ----------------------------

@router.post(
    "",
    response_model=VendorResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_vendor(
    vendor: VendorCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return create_vendor_service(db, vendor)


@router.get(
    "",
    response_model=list[VendorResponse],
)
def get_all_vendors(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_all_vendors_service(db)


@router.get(
    "/{vendor_id}",
    response_model=VendorResponse,
)
def get_vendor_by_id(
    vendor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_vendor_by_id_service(db, vendor_id)


@router.put(
    "/{vendor_id}",
    response_model=VendorResponse,
)
def update_vendor(
    vendor_id: int,
    vendor: VendorUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return update_vendor_service(db, vendor_id, vendor)


@router.delete(
    "/{vendor_id}",
)
def delete_vendor(
    vendor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return delete_vendor_service(db, vendor_id)


# ----------------------------
# Vendor Approval Workflow APIs
# ----------------------------

@router.get(
    "/pending",
    response_model=list[VendorResponse],
)
def get_pending_vendors(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_pending_vendors_service(db)


@router.put(
    "/{vendor_id}/approve",
    response_model=VendorApprovalResponse,
)
def approve_vendor(
    vendor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return approve_vendor_service(db, vendor_id)


@router.put(
    "/{vendor_id}/reject",
    response_model=VendorApprovalResponse,
)
def reject_vendor(
    vendor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return reject_vendor_service(db, vendor_id)

File: frontend/src/app/core/services/vendor.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Vendor } from '../models/vendor.model';

@Injectable({
  providedIn: 'root'
})
export class VendorService {

  private http = inject(HttpClient);

  private readonly API_URL = 'http://127.0.0.1:8000/vendors';

  getAllVendors(): Observable<Vendor[]> {
    return this.http.get<Vendor[]>(this.API_URL);
  }

  getVendorById(vendorId: number): Observable<Vendor> {
    return this.http.get<Vendor>(
      `${this.API_URL}/${vendorId}`
    );
  }

  createVendor(vendor: Vendor): Observable<Vendor> {
    return this.http.post<Vendor>(
      this.API_URL,
      vendor
    );
  }

  updateVendor(
    vendorId: number,
    vendor: Vendor
  ): Observable<Vendor> {
    return this.http.put<Vendor>(
      `${this.API_URL}/${vendorId}`,
      vendor
    );
  }

  deleteVendor(vendorId: number): Observable<any> {
    return this.http.delete(
      `${this.API_URL}/${vendorId}`
    );
  }

  getPendingVendors(): Observable<Vendor[]> {
    return this.http.get<Vendor[]>(
      `${this.API_URL}/pending`
    );
  }

  approveVendor(vendorId: number): Observable<any> {
    return this.http.put(
      `${this.API_URL}/${vendorId}/approve`,
      {}
    );
  }

  rejectVendor(vendorId: number): Observable<any> {
    return this.http.put(
      `${this.API_URL}/${vendorId}/reject`,
      {}
    );
  }
}
